run.c

```diff
@@ -73,43 +73,51 @@ void run(void)
 
         int is_PM = is_PM_timestep(All.Ti_Current);
 
-        enum ActionType action = human_interaction(is_PM, lastPM, TimeLastOutput);
-        switch(action) {
-            case STOP:
-                message(0, "human controlled stop with checkpoint at next PM.\n");
-                /*Write when the PM timestep completes*/
-                Ti_nextoutput = All.Ti_Current;
-                /* next loop will write a new snapshot file; break is for switch */
-                break;
-
-            case TIMEOUT:
-                message(0, "Stopping due to TimeLimitCPU.\n");
-                Ti_nextoutput = All.Ti_Current;
-                break;
-
-            case AUTO_CHECKPOINT:
-                message(0, "Auto checkpoint due to AutoSnapshotTime.\n");
-                Ti_nextoutput = All.Ti_Current;
-                /* will write a new snapshot file next time the PM step finishes*/
-                break;
-
-            case CHECKPOINT:
-                message(0, "human controlled checkpoint at next PM.\n");
-                Ti_nextoutput = All.Ti_Current;
-                /* will write a new snapshot file next time the PM step finishes*/
-                break;
-
-            case TERMINATE:
-                message(0, "human controlled termination.\n");
-                /* no snapshot, at termination, directly end the loop */
-                return;
-
-            case IOCTL:
-            case NO_ACTION:
-                break;
+        SyncPoint * current_sync = find_current_sync_point(All.Ti_Current);
+        next_sync = find_next_sync_point(All.Ti_Current);
+
+        enum ActionType action = NO_ACTION;
+
+        if(is_PM) {
+            action = human_interaction(lastPM, TimeLastOutput);
+            switch(action) {
+                case STOP:
+                    message(0, "human controlled stop with checkpoint at next PM.\n");
+
+                    /* Write when the PM timestep completes*/
+                    current_sync = get_pm_sync_point(All.Ti_Current);
+                    next_sync = NULL;
+                    break;
+
+                case TIMEOUT:
+                    message(0, "Stopping due to TimeLimitCPU.\n");
+                    current_sync = get_pm_sync_point(All.Ti_Current);
+                    next_sync = NULL;
+                    break;
+
+                case AUTO_CHECKPOINT:
+                    message(0, "Auto checkpoint due to AutoSnapshotTime.\n");
+                    current_sync = get_pm_sync_point(All.Ti_Current);
+                    break;
+
+                case CHECKPOINT:
+                    message(0, "human controlled checkpoint at next PM.\n");
+                    current_sync = get_pm_sync_point(All.Ti_Current);
+                    break;
+
+                case TERMINATE:
+                    message(0, "human controlled termination.\n");
+                    /* no snapshot, at termination, directly end the loop */
+                    return;
```

Comment:
Sometimes we do a loop in the jobscript to account for node failures or other strangeness (parameter tweaks, e.g.) within a job. 

I agree this is something at the OS level and we shouldn't have it here. But I worry it is not always possible to send KILL to the process. On Cray there is apkill to send signal to the application. I don't know of such a thing on SGI.

Reply:
Ok, you're the guy who uses it. Just checking it was ok to wait for the PM step for this.

Reply:
Let'sl wait for the response from pfe's support team. If there is a way of terminating the application on SGI too, then we can remove the feature.

Reply:
Response from nasa's support team:

```
Not easily but what you can do is find the node that is the master node for the job.
(qstat -n jobid, the first node listed is the master node)
ssh onto that node and then run the command:
ps axuw|grep mpiexec
then run the command 'kill -2 pid'  where pid is the value from the 2nd column of the ps command
for the main mpiexec process.
if the PBS script has the mpiexec command in loop it will continue on doing whatever it
should do for the next step of the loop.
```
So there is a way and we can get rid of TERMINATE.

Reply:
Or leave it as is -- only terminate on PM step, as the no snapshot counter part of 'stop'.

Reply:
Sure, leave it as it is. It has zero cost - I was only wondering if you wanted to make it terminate really immediately, but that seems like overkill now, I guess.